Rest Client fail on redirected calls when prometheus metrics are configured

### Describe the bug

In a Quarkus project that is configured with prometheus micrometer and quarkus rest client with `follow-redirects` enabled, any redirected call (302) will not complete and but fail with an exception.

In the logs we notice this error:

```
        Suppressed: java.lang.IllegalArgumentException: Prometheus requires that all meters with the same name have the same set of tag keys. There is already an existing meter named 'http_client_requests_seconds' containing tag keys [address, clientName, method, outcome, status, uri]. The meter you are attempting to register has keys [clientName, method, outcome, status, uri].
```

The issue started happening on 3.34.1. The issue was not apparent on 3.33.1

### Expected behavior

Rest Client calls with redirect should complete

### Actual behavior

Rest Client call with redirect fail

### How to Reproduce?

Reproducer found here: https://github.com/pcasaes/quarkus-redirect-302-repro

Follow it's readme

### Output of `uname -a` or `ver`

Darwin XYZ 25.3.0 Darwin Kernel Version 25.3.0: Wed Jan 28 20:51:28 PST 2026; root:xnu-12377.91.3~2/RELEASE_ARM64_T6041 arm64

### Output of `java -version`

openjdk version "21.0.8" 2025-07-15 LTS OpenJDK Runtime Environment Corretto-21.0.8.9.1 (build 21.0.8+9-LTS) OpenJDK 64-Bit Server VM Corretto-21.0.8.9.1 (build 21.0.8+9-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.34.6

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d) Maven home: /Users/xyz/.m2/wrapper/dists/apache-maven-3.8.1-bin/2l5mhf2pq2clrde7f7qp1rdt5m/apache-maven-3.8.1 Java version: 21.0.8, vendor: Amazon.com Inc., runtime: /Users/paulo.casaes/.sdkman/candidates/java/21.0.8-amzn Default locale: en_US, platform encoding: UTF-8 OS name: "mac os x", version: "26.3.1", arch: "aarch64", family: "mac"

### Additional information

Quarkus as of 3.34 will start treating metrics with inconsistent tags as an error where before it was quietly dropping such metrics: https://github.com/quarkusio/quarkus/issues/52410

The issue is that under the described conditions Quarkus is recording the metric twice for the second redirected call but in different code paths and with different tags.

The bug was likely introduced by commit [f5ec75c0cb7](https://github.com/quarkusio/quarkus/commit/f5ec75c0cb7):

* [f5ec75c0cb7](https://github.com/quarkusio/quarkus/commit/f5ec75c0cb7) Ensure that the REST Client registers Vert.x HTTP Client metrics

That commit added:

* [RestClientBuilderMetricsListener](https://github.com/quarkusio/quarkus/commit/f5ec75c0cb7#diff-1a62b588389a4a19b108315c19a75e0ba89d5321008dbe8d8d7eb7c3ed1b1e19)
* [VertxRequestCustomizingClientBuilder](https://github.com/quarkusio/quarkus/commit/f5ec75c0cb7#diff-8715d91079751ec6b6c1c7018af304ba436b8a6ac3a4acbdc9cdf9db8dcba8df)
* the clientRequestCustomizer plumbing through the REST client builder/handler stack


Why it breaks with redirects:

  1. The Micrometer fix suppresses duplicate Vert.x metrics by marking the request metric as ignored here:

* [RestClientBuilderMetricsListener.java:24](https://github.com/quarkusio/quarkus/commit/f5ec75c0cb7#diff-1a62b588389a4a19b108315c19a75e0ba89d5321008dbe8d8d7eb7c3ed1b1e19R24)


  2. That suppression is applied only when Quarkus gets the initial HttpClientRequest back from:

* [ClientSendRequestHandler.java:108](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java#L108)
* customizers run at [ClientSendRequestHandler.java:119](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java#L119)


  3. Redirects are delegated to Vert.x internally via:

* [ClientSendRequestHandler.java:473](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java#L473)
* it sets RequestOptions.setFollowRedirects(followRedirects)
* then calls httpClient.request(options, handler) at [ClientSendRequestHandler.java:482](https://github.com/quarkusio/quarkus/blob/main/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java#482)

  4. **The redirected follow-up request is then created by Vert.x, not by Quarkus code, so Quarkus never reruns the httpClientRequestCustomizer for that second request.**

Result:

* first request may be ignored on the Vert.x side
* redirected request is not ignored
* REST client metrics still run
* Vert.x client metrics also run for the redirected request
* Prometheus sees same meter name with different tag sets and throws

EDIT: In my original application the symptom was a timeout but in the simpler reproducer its a simple exception

I have the same issue when I use a grpc client: 

```
2026-05-12 12:37:48,408 ERROR [org.jboss.resteasy.reactive.server.handlers.PublisherResponseHandler] (vert.x-eventloop-thread-2) Exception in SSE server handling, impossible to send it to client: io.grpc.StatusRuntimeException: UNKNOWN
	at io.grpc.Status.asRuntimeException(Status.java:532)
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:581)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.quarkus.grpc.runtime.supports.IOThreadClientInterceptor$1$1.lambda$onClose$3(IOThreadClientInterceptor.java:70)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.impl.ContextInternal.lambda$runOnContext$0(ContextInternal.java:50)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:566)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.IllegalArgumentException: Prometheus requires that all meters with the same name have the same set of tag keys. There is already an existing meter named 'http_client_requests_seconds' containing tag keys [clientName, method, outcome, status, uri]. The meter you are attempting to register has keys [address, clientName, method, outcome, status, uri].
	at io.micrometer.prometheus.PrometheusMeterRegistry.lambda$throwExceptionOnRegistrationFailure$19(PrometheusMeterRegistry.java:619)
	at io.micrometer.core.instrument.MeterRegistry.meterRegistrationFailed(MeterRegistry.java:1291)
	at io.micrometer.prometheus.PrometheusMeterRegistry.lambda$applyToCollector$18(PrometheusMeterRegistry.java:593)
	at java.base/java.util.concurrent.ConcurrentHashMap.compute(ConcurrentHashMap.java:1956)
	at io.micrometer.prometheus.PrometheusMeterRegistry.applyToCollector(PrometheusMeterRegistry.java:579)
	at io.micrometer.prometheus.PrometheusMeterRegistry.newTimer(PrometheusMeterRegistry.java:328)
	at io.micrometer.core.instrument.MeterRegistry.lambda$timer$6(MeterRegistry.java:379)
	at io.micrometer.core.instrument.MeterRegistry.getOrCreateMeter(MeterRegistry.java:725)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:652)
	at io.micrometer.core.instrument.MeterRegistry.timer(MeterRegistry.java:377)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:471)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:465)
	at io.micrometer.core.instrument.composite.CompositeTimer.registerNewMeter(CompositeTimer.java:205)
	at io.micrometer.core.instrument.composite.CompositeTimer.registerNewMeter(CompositeTimer.java:35)
	at io.micrometer.core.instrument.composite.AbstractCompositeMeter.add(AbstractCompositeMeter.java:67)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at java.base/java.util.Collections$SetFromMap.forEach(Collections.java:6060)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lambda$new$0(CompositeMeterRegistry.java:67)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lock(CompositeMeterRegistry.java:189)
	at io.micrometer.core.instrument.composite.CompositeMeterRegistry.lambda$new$1(CompositeMeterRegistry.java:67)
	at io.micrometer.core.instrument.MeterRegistry.getOrCreateMeter(MeterRegistry.java:735)
	at io.micrometer.core.instrument.MeterRegistry.registerMeterIfNecessary(MeterRegistry.java:652)
	at io.micrometer.core.instrument.MeterRegistry.timer(MeterRegistry.java:377)
	at io.micrometer.core.instrument.Timer$Builder.register(Timer.java:471)
	at io.micrometer.core.instrument.Timer$Builder.lambda$withRegistry$0(Timer.java:453)
	at io.quarkus.micrometer.runtime.binder.vertx.VertxHttpClientMetrics$3.responseEnd(VertxHttpClientMetrics.java:195)
	at io.quarkus.micrometer.runtime.binder.vertx.VertxHttpClientMetrics$3.responseEnd(VertxHttpClientMetrics.java:108)
	at io.vertx.core.http.impl.Http2ClientConnection.metricsEnd(Http2ClientConnection.java:212)
	at io.vertx.core.http.impl.Http2ClientConnection.access$200(Http2ClientConnection.java:39)
	at io.vertx.core.http.impl.Http2ClientConnection$Stream.onEnd(Http2ClientConnection.java:322)
	at io.vertx.core.http.impl.Http2ClientConnection.onHeadersRead(Http2ClientConnection.java:206)
	at io.vertx.core.http.impl.Http2ConnectionBase.onHeadersRead(Http2ConnectionBase.java:209)
	at io.netty.handler.codec.http2.Http2FrameListenerDecorator.onHeadersRead(Http2FrameListenerDecorator.java:46)
	at io.netty.handler.codec.http2.Http2EmptyDataFrameListener.onHeadersRead(Http2EmptyDataFrameListener.java:63)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:430)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder$FrameReadListener.onHeadersRead(DefaultHttp2ConnectionDecoder.java:345)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:489)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:497)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:257)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:174)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:186)
	at io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:61)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:402)
	at io.netty.handler.codec.http2.Http2ConnectionHandler.decode(Http2ConnectionHandler.java:462)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:545)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:484)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:296)
	at io.vertx.core.http.impl.VertxHttp2ConnectionHandler.channelRead(VertxHttp2ConnectionHandler.java:419)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:171)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	... 4 more
```

--- a/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/RestClientRedirectPrometheusTest.java
+++ b/extensions/micrometer/deployment/src/test/java/io/quarkus/micrometer/deployment/binder/RestClientRedirectPrometheusTest.java
@@ -1,0 +1,99 @@
+package io.quarkus.micrometer.deployment.binder;
+
+import static io.restassured.RestAssured.when;
+import static org.hamcrest.Matchers.equalTo;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.quarkus.micrometer.test.Util;
+import io.quarkus.test.QuarkusExtensionTest;
+
+public class RestClientRedirectPrometheusTest {
+
+    @RegisterExtension
+    static final QuarkusExtensionTest config = new QuarkusExtensionTest()
+            .setFlatClassPath(true)
+            .withConfigurationResource("test-logging.properties")
+            .overrideConfigKey("quarkus.micrometer.binder-enabled-default", "false")
+            .overrideConfigKey("quarkus.micrometer.binder.http-client.enabled", "true")
+            .overrideConfigKey("quarkus.micrometer.binder.http-server.enabled", "true")
+            .overrideConfigKey("quarkus.micrometer.binder.vertx.enabled", "true")
+            .overrideConfigKey("quarkus.micrometer.export.prometheus.enabled", "true")
+            .overrideConfigKey("quarkus.micrometer.registry-enabled-default", "false")
+            .overrideConfigKey("quarkus.redis.devservices.enabled", "false")
+            .overrideConfigKey("quarkus.rest-client.redirect-client.url", "${test.url}")
+            .overrideConfigKey("quarkus.rest-client.redirect-client.follow-redirects", "true")
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(Util.class, RedirectResource.class, RedirectResource.RedirectClient.class));
+
+    @Inject
+    MeterRegistry registry;
+
+    @Test
+    public void redirectShouldNotRegisterConflictingClientMeters() throws InterruptedException {
+        when()
+                .get("/exercise")
+                .then()
+                .statusCode(200)
+                .body(equalTo("Hello World!"));
+
+        Util.waitForMeters(registry.find("http.client.requests").timers(), 1);
+        Assertions.assertEquals(1, registry.find("http.client.requests").timers().size(),
+                Util.foundClientRequests(registry, "Expected a single REST client timer after following a redirect."));
+    }
+
+    @Path("/")
+    @Singleton
+    public static class RedirectResource {
+
+        @RegisterRestClient(configKey = "redirect-client")
+        public interface RedirectClient {
+
+            @GET
+            @Path("/redirect")
+            @Produces(MediaType.TEXT_PLAIN)
+            Response getThroughRedirect();
+        }
+
+        @Inject
+        @RestClient
+        RedirectClient redirectClient;
+
+        @GET
+        @Path("/redirect")
+        public Response redirect() {
+            return Response.status(302)
+                    .header("Location", "/target")
+                    .build();
+        }
+
+        @GET
+        @Path("/target")
+        @Produces(MediaType.TEXT_PLAIN)
+        public String target() {
+            return "Hello World!";
+        }
+
+        @GET
+        @Path("/exercise")
+        @Produces(MediaType.TEXT_PLAIN)
+        public String exercise() {
+            try (Response response = redirectClient.getThroughRedirect()) {
+                return response.readEntity(String.class);
+            }
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java
+++ b/independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java
@@ -116,11 +116,7 @@ public class ClientSendRequestHandler implements ClientRestHandler {
                     return;
                 }
 
-                for (int i = 0; i < clientRequestCustomizers.size(); i++) {
-                    clientRequestCustomizers.get(i).accept(httpClientRequest);
-                }
-
-                requestContext.setHttpClientRequest(httpClientRequest);
+                customizeRequest(httpClientRequest, requestContext);
 
                 // adapt headers to HTTP/2 depending on the underlying HTTP connection
                 ClientSendRequestHandler.this.adaptRequest(httpClientRequest);
@@ -254,6 +250,71 @@ public class ClientSendRequestHandler implements ClientRestHandler {
                     reportFinish(event, requestContext);
                 }
             }
+        });
+    }
+
+    private void customizeRequest(HttpClientRequest httpClientRequest, RestClientRequestContext requestContext) {
+        installRedirectRequestCustomizer(httpClientRequest, requestContext);
+        for (int i = 0; i < clientRequestCustomizers.size(); i++) {
+            clientRequestCustomizers.get(i).accept(httpClientRequest);
+        }
+
+        requestContext.setHttpClientRequest(httpClientRequest);
+    }
+
+    /*
+     * If the request leads to a new request because of redirect then this method will install the original request's
+     * customizers to the new redirect request.
+     */
+    private void installRedirectRequestCustomizer(HttpClientRequest httpClientRequest,
+            RestClientRequestContext requestContext) {
+        if (!followRedirects || clientRequestCustomizers.isEmpty()) {
+            return;
+        }
+        httpClientRequest.redirectHandler(response -> {
+            /*
+             * The docs for `redirectHandler()` make no statement about whether the returned result is nullable.
+             * We'll guard against null here and assume there's no redirect taking place.
+             */
+            Function<HttpClientResponse, Future<RequestOptions>> redirectHandler = requestContext.getHttpClient()
+                    // For Vert.x 5 we'll have to cast the `HttpClient` to `HttpClientInternal` in order to access
+                    // this method
+                    .redirectHandler();
+            if (redirectHandler == null) {
+                return Future.succeededFuture(null);
+            }
+
+            /*
+             * @see HttpClient#redirectHandler(Function)
+             * According to the docs the `redirectHandler` function returns null if there is no redirect taking place.
+             *
+             * But it doesn't return `HttpClientRequest`, it returns a `Future<RequestOptions>`.
+             * Vert.x 3 exposed redirect handling in terms of HttpClientRequest, while Vert.x 4
+             * changed the client-level redirect handler to Future<RequestOptions>. The current
+             * docs still describe the old null semantics, so they are likely stale.
+             *
+             * In practice, we defensively accept both interpretations of "no redirect":
+             * - the handler returns a null Future<RequestOptions>
+             * - the handler returns a Future completing with null RequestOptions
+             */
+            Future<RequestOptions> redirectOptions = redirectHandler.apply(response);
+            if (redirectOptions == null) {
+                return Future.succeededFuture(null);
+            }
+            return redirectOptions.compose(options -> {
+                if (options == null) {
+                    return Future.succeededFuture(null);
+                }
+                /*
+                 * If there is a redirect taking place install the original request's customizers to the redirect
+                 * request.
+                 */
+                return requestContext.getHttpClient().request(options)
+                        .map(nextRequest -> {
+                            customizeRequest(nextRequest, requestContext);
+                            return nextRequest;
+                        });
+            });
         });
     }
 

